plugins/userprofileservice/services/in_memory_ups.go

```diff
@@ -0,0 +1,124 @@
+/****************************************************************************
+ * Copyright 2022, Optimizely, Inc. and contributors                        *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package services //
+package services
+
+import (
+	"sync"
+
+	"github.com/optimizely/agent/plugins/userprofileservice"
+	"github.com/optimizely/go-sdk/pkg/decision"
+)
+
+const userIDKey = "user_id"
+const experimentBucketMapKey = "experiment_bucket_map"
+
+// InMemoryUserProfileService represents the in-memory implementation of UserProfileService interface
+type InMemoryUserProfileService struct {
+	Capacity int `json:"capacity"`
+	// StorageStrategy defines the storage strategy. Supported values include fifo and lifo.
+	StorageStrategy     string `json:"storageStrategy"`
+	ProfilesMap         map[string]decision.UserProfile
+	fifoOrderedProfiles chan string
+	lifoOrderedProfiles []string
+	lock                sync.RWMutex
+	isReady             bool
```

Comment:
why we need isReady?

Reply:
since we need to initialize `fifoOrderedProfiles` with `capacity` and this can only be done when first save is called, `isReady` is used to this purpose. All other required initialisations also take place if `isReady` is false.

`u.fifoOrderedProfiles = make(chan string, u.Capacity)`